### DISCUSSION TITLE
What cue module (and directory) structure should I generate for my users?

### DISCUSSION THREAD

**Discussion by dontlaugh** [cue-lang/cue#1282] [Q&A]
I would like to build a tool that can delegate to `cue mod init`, but also pre-populate some schemas and constraints.

Where should these schemas live?

Here?
```
cue.mod/
    gen/
        myschemas.cue
```

or here (outside the module folder)?

```
prebuiltschemas/
    myschemas.cue
cue.mod/
```

Adding them inside the module folder is - I suppose - more aesthetically pleasing, but your docs say that the module folder should be managed by the `cue` tool.

Is there an alternative I'm not seeing for giving my users access to my pre-defined schemas?

**Comment by dontlaugh** [comment]
Update: I had the same thought process as this user https://github.com/cue-lang/cue/issues/1261#issuecomment-921420708

It feels okay to put my own (generated) schemas in gen/, but I figured I'd ask here :)

**Comment by verdverm** [comment]
I think the answer will depend on how you want them to import the prebuilt schemas.

I'd shy away from the `cue.mod` directory. Generally this directory should be considered to be managed by the `cue` tool. The `gen` directory is certainly better than `pkg` if you go this route. If you do use anything under `cue.mod` you will need at least a domain prefix. This could complicate things down the line when CUE has dependency management.

If they should be local to the user module, you should put them outside of `cue.mod`.

**Comment by joulaud** [comment]
I am banging my head around the "packages" and "import" notions in CUE.

I don't really understand several things on what seems an underdocumented corner (even language reference seems really ambiguous and incomplete).

What I did understand at this point:
- you should always work in a module
- having files for two different packages in same directory is a bad idea (at least until https://github.com/cue-lang/cue/discussions/1237 is sorted out)
- a package is comprised of all files with same `package` declaration in the current module
- an instance is all file from the same package from current directory and all its parent until top-module directory
- `cue export` and `cue eval` work on instances
- you can import builtin packages easily
- you should be able to import other packages from current module
- you can import other packages if you put them in some magical directories under `cue.mod`

What I am struggling to understand:
- how does cue resolve an import name?
- in a module `example.com/module`, is it necessary to put the package `example.com/module/a` in directory `/a`?
- should the package declaration be `package a` or `example.com/module/a`?
- what does `cue eval dir/...` exactly means? (it seems to me that `dir/...` is the list of all possible instances of current package in all subdirectories of `dir`)
- when using `cue eval path/to/file.cue` is it true that it only uses the file passed as argument with no regards for any package declaration?

And things I have no idea how it works:
- what is the use of subdirectories (`pkg`, `gen`, `usr`) of `cue.mod`? docs says it is handled by `cue` itself and should not be touched by user but I cannot see a way to tell it to put something in other thing than `gen`
- what is the recommended way to split a configuration ?

I have not read the doc extensively so I could have missed some evident explanation for all of this but if anyone can help me on this I would be grateful.

**Reply by eonpatapon** [reply]
> What I am struggling to understand:
> 
>     * how does cue resolve an import name?

Currently cue doesn't have package management builtin (see https://github.com/cue-lang/cue/issues/851). But you can manually fetch cue modules and put them in `cue.mod/pkg`.

Modules are resolved much like Go. If you want to import `example.com/x`, cue expects to find a similar path in `cue.mod/{pkg,usr,gen}` with cue files inside.

> 
>     * in a module `example.com/module`, is it necessary to put the package `example.com/module/a` in directory `/a`?
> 

No. But if the name of the directory is different from the package name declared in cue file you need to be explicit: 

```cue
import "example.com/module/xxx:a"
```

Where the `xxx` directory contains cue files belonging to package `a`.

>     * should the package declaration be `package a` or `example.com/module/a`?

`package a`
